packages/artifact/src/internal/upload-http-client.ts

```diff
@@ -248,91 +248,86 @@ export class UploadHttpClient {
-          while (offset < uploadFileSize) {
-            const chunkSize = Math.min(
-              uploadFileSize - offset,
-              parameters.maxChunkSize
-            )
+        const start = offset
+        const end = offset + chunkSize - 1
+        offset += parameters.maxChunkSize
 
-            // if an individual file is greater than 100MB (1024*1024*100) in size, display extra information about the upload status
-            if (uploadFileSize > 104857600) {
-              this.statusReporter.updateLargeFileStatus(
-                parameters.file,
-                offset,
-                uploadFileSize
-              )
-            }
+        if (abortFileUpload) {
+          // if we don't want to continue in the event of an error, any pending upload chunks will be marked as failed
+          failedChunkSizes += chunkSize
+          continue
+        }
 
-            const start = offset
-            const end = offset + chunkSize - 1
-            offset += parameters.maxChunkSize
+        const result = await this.uploadChunk(
+          httpClientIndex,
+          parameters.resourceUrl,
+          fs.createReadStream(uploadFilePath, {
+            start,
+            end,
+            autoClose: false
+          }),
+          start,
+          end,
+          uploadFileSize,
+          isGzip,
+          totalFileSize
+        )
 
-            if (abortFileUpload) {
-              // if we don't want to continue in the event of an error, any pending upload chunks will be marked as failed
-              failedChunkSizes += chunkSize
-              continue
-            }
+        if (!result) {
+          // Chunk failed to upload, report as failed and do not continue uploading any more chunks for the file. It is possible that part of a chunk was
+          // successfully uploaded so the server may report a different size for what was uploaded
+          isUploadSuccessful = false
+          failedChunkSizes += chunkSize
+          core.warning(`Aborting upload for ${parameters.file} due to failure`)
+          abortFileUpload = true
+        }
+      }
 
-            const result = await this.uploadChunk(
-              httpClientIndex,
-              parameters.resourceUrl,
-              fs.createReadStream(uploadFilePath, {
-                start,
-                end,
-                autoClose: false
-              }),
-              start,
-              end,
-              uploadFileSize,
-              isGzip,
-              totalFileSize
-            )
+      // Delete the temporary file that was created as part of the upload. Even though this operation is async, we don't need to
+      // await this at it unnecessarily could block other uploads from starting faster. If the temp file does not get manually deleted by
+      // calling cleanup, it gets removed when the node process exits. For more info see: https://www.npmjs.com/package/tmp-promise#about
+      tempFile.cleanup()
```

Comment:
Most important line in this PR is here. We could call `await ` since it's async but I don't see a point since it just prevents other uploads from potentially starting faster

Reply:
I think we should `await` this, especially with the recent concerns around disk size it seems like it would be better to clean up as we go. Also, I think it's just good practice to `await` any async code.

Do you see any noticeable performance differences when you *don't* await this?

Reply:
`await` shouldn't block right?  The thread will just put this on the event loop and move on to the next upload?  I also think we will spend most of the time waiting on network transferring, so my guess is this won't cause measurable delays.

Reply:
Await will block. I did this because there actually is a measurable delay, check out this small snippet
```
import * as tmp from 'tmp-promise'
import {performance} from 'perf_hooks'

let startTime1
let startTime2

export async function test(): Promise<void> {
    await mockCleanupNoAwait()
    var endTime1 = performance.now()
    console.log("Call with no await " + (endTime1 - startTime1) + " milliseconds.")

    await mockCleanupAwait()
    var endTime2 = performance.now()
    console.log("Call with await " + (endTime2 - startTime2) + " milliseconds.")
}

async function mockCleanupNoAwait(): Promise<void>{
    const tempFile = await tmp.file()
    console.log(tempFile)
    startTime1 = performance.now()
    tempFile.cleanup()
    return
}

async function mockCleanupAwait(): Promise<void>{
    const tempFile = await tmp.file()
    console.log(tempFile)
    startTime2 = performance.now()
    await tempFile.cleanup()
    return
}

test()
```

If you run this, this is the result:
![image](https://user-images.githubusercontent.com/16109154/79776564-3a6e4800-8336-11ea-87a0-f7eb65794b6d.png)

I tried a few variations of this and with the `await` it's always slower by a significant margin. For this package this means that the next file upload won't start until the `return` is hit. 

I'm not concerned about storage at all. I think it's better to start the next upload ASAP.

Reply:
I don't think that timing is taking into account the time spent by node later cleaning up the temp file. I would expect the overall perf of an artifact upload to not differ much.

Reply:
The timing doesn't take into account when it was actually deleted, it's only measuring when the promise returns and resolves which is the point. We don't start uploading another file until we return from the previous upload. The reasoning behind not having the await is to let the cleanup finish behind the scenes without it blocking the start of the next upload. 

Regardless of the actual time implications, I added back the await.

Reply:
Well, this test is a simplified version of what's really going on right?  :) In this test it does block because there is nothing else to do.  However in the real action, we have plenty stuff to do all the time, e.g. write logs, upload chunks, and other async code.  

`async / await` by definition means don't block, but do it in background right?  So if you got stuff to do, it won't increase any time, if you don't, then yeah, you will wait for it to finish.

Reply:
Yeah the test is a little simplified. I did a larger large scale test locally and it's pretty hard to draw any conclusions as it starts to vary significantly once I start adding `await Promise.All` and a bunch of other things that this package actually does 🤷‍♂️